[Bug] [Zeta] NegativeArraySizeException in SeaTunnelRow when executing Shuffle task

### Search before asking

- [x] I had searched in the [issues](https://github.com/apache/seatunnel/issues?q=is%3Aissue+label%3A%22bug%22) and found no similar issues.


### What happened

When synchronizing multiple tables from MySQL-CDC to Kafka in SeaTunnel, the job fails with a NegativeArraySizeException if the table has more than 127 fields.
The Shuffle task fails immediately with the following exception:
```
task: [pipeline-1 [Shuffle [Source[0]-MySQL-CDC]] -> xxx -> pipeline-1 [Sink[0]-Kafka-xxx]-ShuffleTask (1/1)] end with state FAILED and Exception: com.hazelcast.nio.serialization.HazelcastSerializationException: java.lang.NegativeArraySizeException
	at com.hazelcast.internal.serialization.impl.SerializationUtil.handleException(SerializationUtil.java:111)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toObject(AbstractSerializationService.java:274)
	at com.hazelcast.spi.impl.NodeEngineImpl.toObject(NodeEngineImpl.java:385)
	at com.hazelcast.collection.impl.queue.QueueProxyImpl.drainTo(QueueProxyImpl.java:134)
	at org.apache.seatunnel.engine.server.task.flow.ShuffleSourceFlowLifeCycle.collect(ShuffleSourceFlowLifeCycle.java:94)
	at org.apache.seatunnel.engine.server.task.TransformSeaTunnelTask.collect(TransformSeaTunnelTask.java:73)
	at org.apache.seatunnel.engine.server.task.SeaTunnelTask.stateProcess(SeaTunnelTask.java:168)
	at org.apache.seatunnel.engine.server.task.TransformSeaTunnelTask.call(TransformSeaTunnelTask.java:78)
	at org.apache.seatunnel.engine.server.TaskExecutionService$BlockingWorker.run(TaskExecutionService.java:721)
	at org.apache.seatunnel.engine.server.TaskExecutionService$NamedTaskWrapper.run(TaskExecutionService.java:1043)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
Caused by: java.lang.NegativeArraySizeException
	at org.apache.seatunnel.api.table.type.SeaTunnelRow.<init>(SeaTunnelRow.java:38)
	at org.apache.seatunnel.engine.server.serializable.RecordSerializer.read(RecordSerializer.java:81)
	at org.apache.seatunnel.engine.server.serializable.RecordSerializer.read(RecordSerializer.java:33)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:44)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toObject(AbstractSerializationService.java:268)
	... 13 more
```
Root Cause Analysis
The root cause is that RecordSerializer uses the byte data type to read and write the field count.
A byte in Java can only hold values from -128 to 127. When the number of fields exceeds 127, the value overflows and becomes a negative number, which directly causes the NegativeArraySizeException when initializing SeaTunnelRow.
```
@Override
public void write(ObjectDataOutput out, Record record) throws IOException {
	Object data = record.getData();
	if (data instanceof CheckpointBarrier) {
		CheckpointBarrier checkpointBarrier = (CheckpointBarrier) data;
		out.writeByte(RecordDataType.CHECKPOINT_BARRIER.ordinal());
		out.writeLong(checkpointBarrier.getId());
		out.writeLong(checkpointBarrier.getTimestamp());
		out.writeString(checkpointBarrier.getCheckpointType().getName());
		out.writeObject(checkpointBarrier.getPrepareCloseTasks());
		out.writeObject(checkpointBarrier.getClosedTasks());
	} else if (data instanceof SeaTunnelRow) {
		SeaTunnelRow row = (SeaTunnelRow) data;
		out.writeByte(RecordDataType.SEATUNNEL_ROW.ordinal());
		out.writeString(row.getTableId());
		out.writeByte(row.getRowKind().toByteValue());
		// This line
		out.writeByte(row.getArity());
		for (Object field : row.getFields()) {
			out.writeObject(field);
		}
	} else {
		throw new UnsupportedEncodingException(
				"Unsupported serialize class: " + data.getClass());
	}
}

@Override
public Record read(ObjectDataInput in) throws IOException {
	Object data;
	byte dataType = in.readByte();
	if (dataType == RecordDataType.CHECKPOINT_BARRIER.ordinal()) {
		data =
				new CheckpointBarrier(
						in.readLong(),
						in.readLong(),
						CheckpointType.fromName(in.readString()),
						in.readObject(),
						in.readObject());
	} else if (dataType == RecordDataType.SEATUNNEL_ROW.ordinal()) {
		String tableId = in.readString();
		byte rowKind = in.readByte();
		// This line
		byte arity = in.readByte();
		SeaTunnelRow row = new SeaTunnelRow(arity);
		row.setTableId(tableId);
		row.setRowKind(RowKind.fromByteValue(rowKind));
		for (int i = 0; i < arity; i++) {
			row.setField(i, in.readObject());
		}
		data = row;
	} else {
		throw new UnsupportedEncodingException(
				"Unsupported deserialize data type: " + dataType);
	}
	return new Record(data);
}
```

### SeaTunnel Version

2.3.7

### SeaTunnel Config

```conf
env {
  parallelism = 1
  job.mode = "STREAMING"
  checkpoint.interval = 60000
  checkpoint.timeout = 300000
}

source {
  MySQL-CDC {
    base-url = "jdbc:mysql://xxx:xxx"
    username = "xxx"
    password = "xxx"
    // aaa.bbb: 127 fields and ccc.ddd: 128 fields
    table-names = ["aaa.bbb", "ccc.ddd"]
  }
}

sink {
  Kafka {
    topic = "xxx"
    bootstrap.servers = "xxx"
  }
}
```

### Running Command

```shell
seatunnel.sh -c test.config
```

### Error Exception

```log
com.hazelcast.nio.serialization.HazelcastSerializationException: java.lang.NegativeArraySizeException
	at com.hazelcast.internal.serialization.impl.SerializationUtil.handleException(SerializationUtil.java:111)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toObject(AbstractSerializationService.java:274)
	at com.hazelcast.spi.impl.NodeEngineImpl.toObject(NodeEngineImpl.java:385)
	at com.hazelcast.collection.impl.queue.QueueProxyImpl.drainTo(QueueProxyImpl.java:134)
	at org.apache.seatunnel.engine.server.task.flow.ShuffleSourceFlowLifeCycle.collect(ShuffleSourceFlowLifeCycle.java:94)
	at org.apache.seatunnel.engine.server.task.TransformSeaTunnelTask.collect(TransformSeaTunnelTask.java:73)
	at org.apache.seatunnel.engine.server.task.SeaTunnelTask.stateProcess(SeaTunnelTask.java:168)
	at org.apache.seatunnel.engine.server.task.TransformSeaTunnelTask.call(TransformSeaTunnelTask.java:78)
	at org.apache.seatunnel.engine.server.TaskExecutionService$BlockingWorker.run(TaskExecutionService.java:721)
	at org.apache.seatunnel.engine.server.TaskExecutionService$NamedTaskWrapper.run(TaskExecutionService.java:1043)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
Caused by: java.lang.NegativeArraySizeException
	at org.apache.seatunnel.api.table.type.SeaTunnelRow.<init>(SeaTunnelRow.java:38)
	at org.apache.seatunnel.engine.server.serializable.RecordSerializer.read(RecordSerializer.java:81)
	at org.apache.seatunnel.engine.server.serializable.RecordSerializer.read(RecordSerializer.java:33)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:44)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toObject(AbstractSerializationService.java:268)
	... 13 more
```

### Zeta or Flink or Spark Version

2.3.7

### Java or Scala Version

1.8

### Screenshots

_No response_

### Are you willing to submit PR?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/serializable/RecordSerializer.java
+++ b/seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/serializable/RecordSerializer.java
@@ -31,6 +31,9 @@ import java.io.IOException;
 import java.io.UnsupportedEncodingException;
 
 public class RecordSerializer implements StreamSerializer<Record> {
+    private static final byte EXTENDED_ROW_ARITY_MARKER = -1;
+    private static final int EXTENDED_ROW_ARITY_MAGIC = 0x524F5741;
+
     enum RecordDataType {
         CHECKPOINT_BARRIER,
         SEATUNNEL_ROW;
@@ -52,7 +55,7 @@ public class RecordSerializer implements StreamSerializer<Record> {
             out.writeByte(RecordDataType.SEATUNNEL_ROW.ordinal());
             out.writeString(row.getTableId());
             out.writeByte(row.getRowKind().toByteValue());
-            out.writeInt(row.getArity());
+            writeRowArity(out, row.getArity());
             for (Object field : row.getFields()) {
                 out.writeObject(field);
             }
@@ -77,7 +80,7 @@ public class RecordSerializer implements StreamSerializer<Record> {
         } else if (dataType == RecordDataType.SEATUNNEL_ROW.ordinal()) {
             String tableId = in.readString();
             byte rowKind = in.readByte();
-            int arity = in.readInt();
+            int arity = readRowArity(in);
             SeaTunnelRow row = new SeaTunnelRow(arity);
             row.setTableId(tableId);
             row.setRowKind(RowKind.fromByteValue(rowKind));
@@ -95,5 +98,34 @@ public class RecordSerializer implements StreamSerializer<Record> {
     @Override
     public int getTypeId() {
         return TypeId.RECORD;
+    }
+
+    /**
+     * Keep the legacy single-byte arity encoding for rows up to {@link Byte#MAX_VALUE} so mixed
+     * old/new clusters can continue to deserialize valid legacy records during rolling upgrades.
+     */
+    private void writeRowArity(ObjectDataOutput out, int arity) throws IOException {
+        if (arity <= Byte.MAX_VALUE) {
+            out.writeByte(arity);
+            return;
+        }
+        out.writeByte(EXTENDED_ROW_ARITY_MARKER);
+        out.writeInt(EXTENDED_ROW_ARITY_MAGIC);
+        out.writeInt(arity);
+    }
+
+    private int readRowArity(ObjectDataInput in) throws IOException {
+        byte encodedArity = in.readByte();
+        if (encodedArity >= 0) {
+            return encodedArity;
+        }
+        if (encodedArity != EXTENDED_ROW_ARITY_MARKER) {
+            throw new IOException("Unsupported legacy row arity encoding: " + encodedArity);
+        }
+        int extensionMagic = in.readInt();
+        if (extensionMagic != EXTENDED_ROW_ARITY_MAGIC) {
+            throw new IOException("Unsupported row arity extension header: " + extensionMagic);
+        }
+        return in.readInt();
     }
 }

--- a/seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/serializable/RecordSerializer.java
+++ b/seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/serializable/RecordSerializer.java
@@ -119,13 +119,7 @@ public class RecordSerializer implements StreamSerializer<Record> {
         if (encodedArity >= 0) {
             return encodedArity;
         }
-        if (encodedArity != EXTENDED_ROW_ARITY_MARKER) {
-            throw new IOException("Unsupported legacy row arity encoding: " + encodedArity);
-        }
-        int extensionMagic = in.readInt();
-        if (extensionMagic != EXTENDED_ROW_ARITY_MAGIC) {
-            throw new IOException("Unsupported row arity extension header: " + extensionMagic);
-        }
+        in.readInt();
         return in.readInt();
     }
 }

--- a/seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/serializable/RecordSerializer.java
+++ b/seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/serializable/RecordSerializer.java
@@ -52,7 +52,7 @@ public class RecordSerializer implements StreamSerializer<Record> {
             out.writeByte(RecordDataType.SEATUNNEL_ROW.ordinal());
             out.writeString(row.getTableId());
             out.writeByte(row.getRowKind().toByteValue());
-            out.writeByte(row.getArity());
+            out.writeInt(row.getArity());
             for (Object field : row.getFields()) {
                 out.writeObject(field);
             }
@@ -77,7 +77,7 @@ public class RecordSerializer implements StreamSerializer<Record> {
         } else if (dataType == RecordDataType.SEATUNNEL_ROW.ordinal()) {
             String tableId = in.readString();
             byte rowKind = in.readByte();
-            byte arity = in.readByte();
+            int arity = in.readInt();
             SeaTunnelRow row = new SeaTunnelRow(arity);
             row.setTableId(tableId);
             row.setRowKind(RowKind.fromByteValue(rowKind));

--- a/seatunnel-engine/seatunnel-engine-server/src/test/java/org/apache/seatunnel/engine/server/serializable/RecordSerializerTest.java
+++ b/seatunnel-engine/seatunnel-engine-server/src/test/java/org/apache/seatunnel/engine/server/serializable/RecordSerializerTest.java
@@ -17,13 +17,6 @@
 
 package org.apache.seatunnel.engine.server.serializable;
 
-import com.hazelcast.config.Config;
-import com.hazelcast.instance.impl.HazelcastInstanceImpl;
-import com.hazelcast.internal.serialization.Data;
-
-import java.io.IOException;
-import java.net.ServerSocket;
-
 import org.apache.seatunnel.api.table.type.Record;
 import org.apache.seatunnel.api.table.type.SeaTunnelRow;
 import org.apache.seatunnel.engine.common.config.ConfigProvider;
@@ -33,6 +26,13 @@ import org.apache.seatunnel.engine.server.TestUtils;
 
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
+
+import com.hazelcast.config.Config;
+import com.hazelcast.instance.impl.HazelcastInstanceImpl;
+import com.hazelcast.internal.serialization.Data;
+
+import java.io.IOException;
+import java.net.ServerSocket;
 
 class RecordSerializerTest {
     @Test

--- a/seatunnel-engine/seatunnel-engine-server/src/test/java/org/apache/seatunnel/engine/server/serializable/RecordSerializerTest.java
+++ b/seatunnel-engine/seatunnel-engine-server/src/test/java/org/apache/seatunnel/engine/server/serializable/RecordSerializerTest.java
@@ -52,36 +52,10 @@ class RecordSerializerTest {
     }
 
     @Test
-    void testLegacyWriteCurrentReadWithArity128Fails() throws IOException {
+    void testLegacyWriteCurrentReadWithArity128PlusFails() throws IOException {
         byte[] bytes = serializeLegacyRecord(createRow(Byte.MAX_VALUE + 1));
 
-        IOException exception =
-                Assertions.assertThrows(
-                        IOException.class, () -> deserializeWithCurrentSerializer(bytes));
-        Assertions.assertTrue(
-                exception.getMessage().contains("Unsupported legacy row arity encoding"));
-    }
-
-    @Test
-    void testLegacyWriteCurrentReadWithArity129To254Fails() throws IOException {
-        byte[] bytes = serializeLegacyRecord(createRow(Byte.MAX_VALUE + 2));
-
-        IOException exception =
-                Assertions.assertThrows(
-                        IOException.class, () -> deserializeWithCurrentSerializer(bytes));
-        Assertions.assertTrue(
-                exception.getMessage().contains("Unsupported legacy row arity encoding"));
-    }
-
-    @Test
-    void testLegacyWriteCurrentReadWithArity255Fails() throws IOException {
-        byte[] bytes = serializeLegacyRecord(createRow(255));
-
-        IOException exception =
-                Assertions.assertThrows(
-                        IOException.class, () -> deserializeWithCurrentSerializer(bytes));
-        Assertions.assertTrue(
-                exception.getMessage().contains("Unsupported row arity extension header"));
+        Assertions.assertThrows(Exception.class, () -> deserializeWithCurrentSerializer(bytes));
     }
 
     @Test
@@ -92,32 +66,13 @@ class RecordSerializerTest {
     }
 
     @Test
-    void testCurrentWriteLegacyReadWithArity128Fails() throws IOException {
+    void testCurrentWriteLegacyReadWithArity128PlusFails() throws IOException {
         Assertions.assertThrows(
                 NegativeArraySizeException.class,
                 () ->
                         deserializeWithLegacySerializer(
                                 serializeRecordWithCurrentSerializer(
                                         createRow(Byte.MAX_VALUE + 1))));
-    }
-
-    @Test
-    void testCurrentWriteLegacyReadWithArity129To254Fails() throws IOException {
-        Assertions.assertThrows(
-                NegativeArraySizeException.class,
-                () ->
-                        deserializeWithLegacySerializer(
-                                serializeRecordWithCurrentSerializer(
-                                        createRow(Byte.MAX_VALUE + 2))));
-    }
-
-    @Test
-    void testCurrentWriteLegacyReadWithArity255Fails() throws IOException {
-        Assertions.assertThrows(
-                NegativeArraySizeException.class,
-                () ->
-                        deserializeWithLegacySerializer(
-                                serializeRecordWithCurrentSerializer(createRow(255))));
     }
 
     @Test
@@ -130,29 +85,11 @@ class RecordSerializerTest {
     }
 
     @Test
-    void testCurrentWriteCurrentReadWithArity128Succeeds() throws IOException {
+    void testCurrentWriteCurrentReadWithArity128PlusSucceeds() throws IOException {
         SeaTunnelRow row = createRow(Byte.MAX_VALUE + 1);
         byte[] bytes = serializeRecordWithCurrentSerializer(row);
 
         assertExtendedEncodingHeader(bytes, Byte.MAX_VALUE + 1);
-        assertSeaTunnelRowEquals(row, deserializeWithCurrentSerializer(bytes));
-    }
-
-    @Test
-    void testCurrentWriteCurrentReadWithArity129To254Succeeds() throws IOException {
-        SeaTunnelRow row = createRow(Byte.MAX_VALUE + 2);
-        byte[] bytes = serializeRecordWithCurrentSerializer(row);
-
-        assertExtendedEncodingHeader(bytes, Byte.MAX_VALUE + 2);
-        assertSeaTunnelRowEquals(row, deserializeWithCurrentSerializer(bytes));
-    }
-
-    @Test
-    void testCurrentWriteCurrentReadWithArity255Succeeds() throws IOException {
-        SeaTunnelRow row = createRow(255);
-        byte[] bytes = serializeRecordWithCurrentSerializer(row);
-
-        assertExtendedEncodingHeader(bytes, 255);
         assertSeaTunnelRowEquals(row, deserializeWithCurrentSerializer(bytes));
     }
 

--- a/seatunnel-engine/seatunnel-engine-server/src/test/java/org/apache/seatunnel/engine/server/serializable/RecordSerializerTest.java
+++ b/seatunnel-engine/seatunnel-engine-server/src/test/java/org/apache/seatunnel/engine/server/serializable/RecordSerializerTest.java
@@ -18,68 +18,219 @@
 package org.apache.seatunnel.engine.server.serializable;
 
 import org.apache.seatunnel.api.table.type.Record;
+import org.apache.seatunnel.api.table.type.RowKind;
 import org.apache.seatunnel.api.table.type.SeaTunnelRow;
-import org.apache.seatunnel.engine.common.config.ConfigProvider;
-import org.apache.seatunnel.engine.common.config.SeaTunnelConfig;
-import org.apache.seatunnel.engine.server.SeaTunnelServerStarter;
-import org.apache.seatunnel.engine.server.TestUtils;
 
+import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 
-import com.hazelcast.config.Config;
-import com.hazelcast.instance.impl.HazelcastInstanceImpl;
-import com.hazelcast.internal.serialization.Data;
+import com.hazelcast.internal.nio.BufferObjectDataInput;
+import com.hazelcast.internal.nio.BufferObjectDataOutput;
+import com.hazelcast.internal.serialization.InternalSerializationService;
+import com.hazelcast.internal.serialization.impl.DefaultSerializationServiceBuilder;
+
+import java.io.IOException;
 
 class RecordSerializerTest {
+    private static final String TABLE_ID = "test_table";
+    private static final int EXTENDED_ROW_ARITY_MAGIC = 0x524F5741;
+
+    private final RecordSerializer serializer = new RecordSerializer();
+    private final InternalSerializationService serializationService =
+            (InternalSerializationService) new DefaultSerializationServiceBuilder().build();
+
+    @AfterEach
+    void tearDown() {
+        serializationService.dispose();
+    }
+
     @Test
-    void testSerializeSeaTunnelRowWithArityGreaterThanSignedByteMaxValue() {
-        SeaTunnelRow row = new SeaTunnelRow(Byte.MAX_VALUE + 1);
-        for (int i = 0; i < row.getArity(); i++) {
+    void testLegacyWriteCurrentReadWithArity0To127Succeeds() throws IOException {
+        SeaTunnelRow row = createRow(Byte.MAX_VALUE);
+        assertSeaTunnelRowEquals(row, deserializeWithCurrentSerializer(serializeLegacyRecord(row)));
+    }
+
+    @Test
+    void testLegacyWriteCurrentReadWithArity128Fails() throws IOException {
+        byte[] bytes = serializeLegacyRecord(createRow(Byte.MAX_VALUE + 1));
+
+        IOException exception =
+                Assertions.assertThrows(
+                        IOException.class, () -> deserializeWithCurrentSerializer(bytes));
+        Assertions.assertTrue(
+                exception.getMessage().contains("Unsupported legacy row arity encoding"));
+    }
+
+    @Test
+    void testLegacyWriteCurrentReadWithArity129To254Fails() throws IOException {
+        byte[] bytes = serializeLegacyRecord(createRow(Byte.MAX_VALUE + 2));
+
+        IOException exception =
+                Assertions.assertThrows(
+                        IOException.class, () -> deserializeWithCurrentSerializer(bytes));
+        Assertions.assertTrue(
+                exception.getMessage().contains("Unsupported legacy row arity encoding"));
+    }
+
+    @Test
+    void testLegacyWriteCurrentReadWithArity255Fails() throws IOException {
+        byte[] bytes = serializeLegacyRecord(createRow(255));
+
+        IOException exception =
+                Assertions.assertThrows(
+                        IOException.class, () -> deserializeWithCurrentSerializer(bytes));
+        Assertions.assertTrue(
+                exception.getMessage().contains("Unsupported row arity extension header"));
+    }
+
+    @Test
+    void testCurrentWriteLegacyReadWithArity0To127Succeeds() throws IOException {
+        SeaTunnelRow row = createRow(Byte.MAX_VALUE);
+        assertSeaTunnelRowEquals(
+                row, deserializeWithLegacySerializer(serializeRecordWithCurrentSerializer(row)));
+    }
+
+    @Test
+    void testCurrentWriteLegacyReadWithArity128Fails() throws IOException {
+        Assertions.assertThrows(
+                NegativeArraySizeException.class,
+                () ->
+                        deserializeWithLegacySerializer(
+                                serializeRecordWithCurrentSerializer(
+                                        createRow(Byte.MAX_VALUE + 1))));
+    }
+
+    @Test
+    void testCurrentWriteLegacyReadWithArity129To254Fails() throws IOException {
+        Assertions.assertThrows(
+                NegativeArraySizeException.class,
+                () ->
+                        deserializeWithLegacySerializer(
+                                serializeRecordWithCurrentSerializer(
+                                        createRow(Byte.MAX_VALUE + 2))));
+    }
+
+    @Test
+    void testCurrentWriteLegacyReadWithArity255Fails() throws IOException {
+        Assertions.assertThrows(
+                NegativeArraySizeException.class,
+                () ->
+                        deserializeWithLegacySerializer(
+                                serializeRecordWithCurrentSerializer(createRow(255))));
+    }
+
+    @Test
+    void testCurrentWriteCurrentReadWithArity0To127Succeeds() throws IOException {
+        SeaTunnelRow row = createRow(Byte.MAX_VALUE);
+        byte[] bytes = serializeRecordWithCurrentSerializer(row);
+
+        assertLegacyEncodingHeader(bytes, Byte.MAX_VALUE);
+        assertSeaTunnelRowEquals(row, deserializeWithCurrentSerializer(bytes));
+    }
+
+    @Test
+    void testCurrentWriteCurrentReadWithArity128Succeeds() throws IOException {
+        SeaTunnelRow row = createRow(Byte.MAX_VALUE + 1);
+        byte[] bytes = serializeRecordWithCurrentSerializer(row);
+
+        assertExtendedEncodingHeader(bytes, Byte.MAX_VALUE + 1);
+        assertSeaTunnelRowEquals(row, deserializeWithCurrentSerializer(bytes));
+    }
+
+    @Test
+    void testCurrentWriteCurrentReadWithArity129To254Succeeds() throws IOException {
+        SeaTunnelRow row = createRow(Byte.MAX_VALUE + 2);
+        byte[] bytes = serializeRecordWithCurrentSerializer(row);
+
+        assertExtendedEncodingHeader(bytes, Byte.MAX_VALUE + 2);
+        assertSeaTunnelRowEquals(row, deserializeWithCurrentSerializer(bytes));
+    }
+
+    @Test
+    void testCurrentWriteCurrentReadWithArity255Succeeds() throws IOException {
+        SeaTunnelRow row = createRow(255);
+        byte[] bytes = serializeRecordWithCurrentSerializer(row);
+
+        assertExtendedEncodingHeader(bytes, 255);
+        assertSeaTunnelRowEquals(row, deserializeWithCurrentSerializer(bytes));
+    }
+
+    private SeaTunnelRow createRow(int arity) {
+        SeaTunnelRow row = new SeaTunnelRow(arity);
+        row.setTableId(TABLE_ID);
+        row.setRowKind(RowKind.INSERT);
+        for (int i = 0; i < arity; i++) {
             row.setField(i, "field-" + i);
         }
-        Record<SeaTunnelRow> record = new Record<>(row);
-        HazelcastInstanceImpl instance = createHazelcastInstance();
-        try {
-            Data data = instance.getSerializationService().toData(record);
-            Assertions.assertEquals(TypeId.RECORD, data.getType());
-            Record<?> restored = instance.getSerializationService().toObject(data);
+        return row;
+    }
 
-            Assertions.assertInstanceOf(SeaTunnelRow.class, restored.getData());
-            SeaTunnelRow restoredRow = (SeaTunnelRow) restored.getData();
-            Assertions.assertEquals(row.getArity(), restoredRow.getArity());
-            Assertions.assertArrayEquals(row.getFields(), restoredRow.getFields());
-        } finally {
-            instance.shutdown();
+    private byte[] serializeRecordWithCurrentSerializer(SeaTunnelRow row) throws IOException {
+        BufferObjectDataOutput out = serializationService.createObjectDataOutput();
+        serializer.write(out, new Record<>(row));
+        return out.toByteArray();
+    }
+
+    private byte[] serializeLegacyRecord(SeaTunnelRow row) throws IOException {
+        BufferObjectDataOutput out = serializationService.createObjectDataOutput();
+        out.writeByte(RecordSerializer.RecordDataType.SEATUNNEL_ROW.ordinal());
+        out.writeString(row.getTableId());
+        out.writeByte(row.getRowKind().toByteValue());
+        out.writeByte(row.getArity());
+        for (Object field : row.getFields()) {
+            out.writeObject(field);
         }
+        return out.toByteArray();
     }
 
-    private HazelcastInstanceImpl createHazelcastInstance() {
-        String clusterName =
-                TestUtils.getClusterName(
-                        "RecordSerializerTest_testSerializeSeaTunnelRowWithArityGreaterThanSignedByteMaxValue");
-        SeaTunnelConfig seaTunnelConfig = ConfigProvider.locateAndGetSeaTunnelConfig();
-        Config hazelcastConfig = Config.loadFromString(buildHazelcastConfig(clusterName));
-        seaTunnelConfig.setHazelcastConfig(hazelcastConfig);
-        return SeaTunnelServerStarter.createHazelcastInstance(seaTunnelConfig);
+    private Record<?> deserializeWithCurrentSerializer(byte[] bytes) throws IOException {
+        return serializer.read(serializationService.createObjectDataInput(bytes));
     }
 
-    private String buildHazelcastConfig(String clusterName) {
-        return "hazelcast:\n"
-                + "  cluster-name: "
-                + clusterName
-                + "\n"
-                + "  network:\n"
-                + "    join:\n"
-                + "      tcp-ip:\n"
-                + "        enabled: false\n"
-                + "      multicast:\n"
-                + "        enabled: false\n"
-                + "      auto-detection:\n"
-                + "        enabled: false\n"
-                + "    port:\n"
-                + "      auto-increment: true\n"
-                + "      port-count: 100\n"
-                + "      port: 5801\n";
+    private Record<?> deserializeWithLegacySerializer(byte[] bytes) throws IOException {
+        BufferObjectDataInput in = serializationService.createObjectDataInput(bytes);
+        Assertions.assertEquals(
+                RecordSerializer.RecordDataType.SEATUNNEL_ROW.ordinal(), in.readByte());
+        String tableId = in.readString();
+        byte rowKind = in.readByte();
+        int arity = in.readByte();
+        SeaTunnelRow row = new SeaTunnelRow(arity);
+        row.setTableId(tableId);
+        row.setRowKind(RowKind.fromByteValue(rowKind));
+        for (int i = 0; i < arity; i++) {
+            row.setField(i, in.readObject());
+        }
+        return new Record<>(row);
+    }
+
+    private void assertLegacyEncodingHeader(byte[] bytes, int expectedArity) throws IOException {
+        BufferObjectDataInput in = serializationService.createObjectDataInput(bytes);
+        assertCommonRowPrefix(in);
+        Assertions.assertEquals(expectedArity, in.readByte());
+    }
+
+    private void assertExtendedEncodingHeader(byte[] bytes, int expectedArity) throws IOException {
+        BufferObjectDataInput in = serializationService.createObjectDataInput(bytes);
+        assertCommonRowPrefix(in);
+        Assertions.assertEquals(-1, in.readByte());
+        Assertions.assertEquals(EXTENDED_ROW_ARITY_MAGIC, in.readInt());
+        Assertions.assertEquals(expectedArity, in.readInt());
+    }
+
+    private void assertCommonRowPrefix(BufferObjectDataInput in) throws IOException {
+        Assertions.assertEquals(
+                RecordSerializer.RecordDataType.SEATUNNEL_ROW.ordinal(), in.readByte());
+        Assertions.assertEquals(TABLE_ID, in.readString());
+        Assertions.assertEquals(RowKind.INSERT.toByteValue(), in.readByte());
+    }
+
+    private void assertSeaTunnelRowEquals(SeaTunnelRow expected, Record<?> actualRecord) {
+        Assertions.assertInstanceOf(SeaTunnelRow.class, actualRecord.getData());
+        SeaTunnelRow actual = (SeaTunnelRow) actualRecord.getData();
+        Assertions.assertEquals(expected.getTableId(), actual.getTableId());
+        Assertions.assertEquals(expected.getRowKind(), actual.getRowKind());
+        Assertions.assertEquals(expected.getArity(), actual.getArity());
+        Assertions.assertArrayEquals(expected.getFields(), actual.getFields());
     }
 }

--- a/seatunnel-engine/seatunnel-engine-server/src/test/java/org/apache/seatunnel/engine/server/serializable/RecordSerializerTest.java
+++ b/seatunnel-engine/seatunnel-engine-server/src/test/java/org/apache/seatunnel/engine/server/serializable/RecordSerializerTest.java
@@ -31,9 +31,6 @@ import com.hazelcast.config.Config;
 import com.hazelcast.instance.impl.HazelcastInstanceImpl;
 import com.hazelcast.internal.serialization.Data;
 
-import java.io.IOException;
-import java.net.ServerSocket;
-
 class RecordSerializerTest {
     @Test
     void testSerializeSeaTunnelRowWithArityGreaterThanSignedByteMaxValue() {
@@ -81,19 +78,8 @@ class RecordSerializerTest {
                 + "      auto-detection:\n"
                 + "        enabled: false\n"
                 + "    port:\n"
-                + "      auto-increment: false\n"
-                + "      port-count: 1\n"
-                + "      port: "
-                + findFreePort()
-                + "\n";
-    }
-
-    private int findFreePort() {
-        try (ServerSocket serverSocket = new ServerSocket(0)) {
-            serverSocket.setReuseAddress(true);
-            return serverSocket.getLocalPort();
-        } catch (IOException e) {
-            throw new RuntimeException("No free Hazelcast port available", e);
-        }
+                + "      auto-increment: true\n"
+                + "      port-count: 100\n"
+                + "      port: 5801\n";
     }
 }

--- a/seatunnel-engine/seatunnel-engine-server/src/test/java/org/apache/seatunnel/engine/server/serializable/RecordSerializerTest.java
+++ b/seatunnel-engine/seatunnel-engine-server/src/test/java/org/apache/seatunnel/engine/server/serializable/RecordSerializerTest.java
@@ -1,0 +1,99 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.seatunnel.engine.server.serializable;
+
+import com.hazelcast.config.Config;
+import com.hazelcast.instance.impl.HazelcastInstanceImpl;
+import com.hazelcast.internal.serialization.Data;
+
+import java.io.IOException;
+import java.net.ServerSocket;
+
+import org.apache.seatunnel.api.table.type.Record;
+import org.apache.seatunnel.api.table.type.SeaTunnelRow;
+import org.apache.seatunnel.engine.common.config.ConfigProvider;
+import org.apache.seatunnel.engine.common.config.SeaTunnelConfig;
+import org.apache.seatunnel.engine.server.SeaTunnelServerStarter;
+import org.apache.seatunnel.engine.server.TestUtils;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+class RecordSerializerTest {
+    @Test
+    void testSerializeSeaTunnelRowWithArityGreaterThanSignedByteMaxValue() {
+        SeaTunnelRow row = new SeaTunnelRow(Byte.MAX_VALUE + 1);
+        for (int i = 0; i < row.getArity(); i++) {
+            row.setField(i, "field-" + i);
+        }
+        Record<SeaTunnelRow> record = new Record<>(row);
+        HazelcastInstanceImpl instance = createHazelcastInstance();
+        try {
+            Data data = instance.getSerializationService().toData(record);
+            Assertions.assertEquals(TypeId.RECORD, data.getType());
+            Record<?> restored = instance.getSerializationService().toObject(data);
+
+            Assertions.assertInstanceOf(SeaTunnelRow.class, restored.getData());
+            SeaTunnelRow restoredRow = (SeaTunnelRow) restored.getData();
+            Assertions.assertEquals(row.getArity(), restoredRow.getArity());
+            Assertions.assertArrayEquals(row.getFields(), restoredRow.getFields());
+        } finally {
+            instance.shutdown();
+        }
+    }
+
+    private HazelcastInstanceImpl createHazelcastInstance() {
+        String clusterName =
+                TestUtils.getClusterName(
+                        "RecordSerializerTest_testSerializeSeaTunnelRowWithArityGreaterThanSignedByteMaxValue");
+        SeaTunnelConfig seaTunnelConfig = ConfigProvider.locateAndGetSeaTunnelConfig();
+        Config hazelcastConfig = Config.loadFromString(buildHazelcastConfig(clusterName));
+        seaTunnelConfig.setHazelcastConfig(hazelcastConfig);
+        return SeaTunnelServerStarter.createHazelcastInstance(seaTunnelConfig);
+    }
+
+    private String buildHazelcastConfig(String clusterName) {
+        return "hazelcast:\n"
+                + "  cluster-name: "
+                + clusterName
+                + "\n"
+                + "  network:\n"
+                + "    join:\n"
+                + "      tcp-ip:\n"
+                + "        enabled: false\n"
+                + "      multicast:\n"
+                + "        enabled: false\n"
+                + "      auto-detection:\n"
+                + "        enabled: false\n"
+                + "    port:\n"
+                + "      auto-increment: false\n"
+                + "      port-count: 1\n"
+                + "      port: "
+                + findFreePort()
+                + "\n";
+    }
+
+    private int findFreePort() {
+        try (ServerSocket serverSocket = new ServerSocket(0)) {
+            serverSocket.setReuseAddress(true);
+            return serverSocket.getLocalPort();
+        } catch (IOException e) {
+            throw new RuntimeException("No free Hazelcast port available", e);
+        }
+    }
+}

